### ISSUE TITLE
Saving and closing file can result in the file getting empty

### ISSUE THREAD

**Issue by waitingsong** [Microsoft/vscode#50942]
Issue Type: <b>Bug</b>

I map `shift+z+z` to save and close active editor with vim plugin.
It works good with vs code version 1.20 ( VSCodeSetup-x64-1.20.1.exe), but got problem with latest version 1.23.1:
1. open a  `.ts` file containing some codes
2. shortcut `shift+z+z` triggered
3. active editor closed
4. the file be zero byte

VS Code version: Code 1.23.1 (d0182c3417d225529c6d5ad24b7572815d0de9ac, 2018-05-10T17:11:17.614Z)
OS version: Windows_NT x64 6.1.7601

<details>
<summary>System Info</summary>

|Item|Value|
|---|---|
|CPUs|Intel(R) Core(TM) i5 CPU         760  @ 2.80GHz (4 x 2809)|
|GPU Status|2d_canvas: enabled<br>flash_3d: enabled<br>flash_stage3d: enabled<br>flash_stage3d_baseline: enabled<br>gpu_compositing: enabled<br>multiple_raster_threads: enabled_on<br>native_gpu_memory_buffers: disabled_software<br>rasterization: unavailable_software<br>video_decode: enabled<br>video_encode: enabled<br>vpx_decode: unavailable_software<br>webgl: enabled<br>webgl2: enabled|
|Memory (System)|15.97GB (5.34GB free)|
|Process Argv|C:\Program Files\Microsoft VS Code\Code.exe D:\tt\tt\lib\index.ts|
|Screen Reader|no|
|VM|67%|

</details><details><summary>Extensions (2)</summary>

Extension|Author (truncated)|Version
---|---|---
vscode-language-pack-zh-hans|MS-|1.23.5
vim|vsc|0.12.0

</details>

config:
```
{
    "editor.fontSize": 18,
    "editor.wordWrap": "on",
    "telemetry.enableTelemetry": false,
    "telemetry.enableCrashReporter": false,
    "files.eol": "\n",
    "files.trimTrailingWhitespace": true,
    "vim.easymotion": false,
    "vim.incsearch": true,
    "vim.useSystemClipboard": true,
    "vim.useCtrlKeys": true,
    "vim.hlsearch": true,
    "vim.insertModeKeyBindings": [
        {
            "before": ["j","j"],
            "after": ["<Esc>"]
        }
    ],
    "vim.otherModesKeyBindingsNonRecursive": [
        {
            "before": ["<leader>","c", "c"],
            "after": ["g", "c"]
        },
        {
            "before":["<C-n>"],
            "after":[],
            "commands": [
                {
                    "command": ":nohl"
                }
            ]
        },
        {
            "before": [
                "Z",
                "Z"
            ],
            "after": [],
            "commands": [
                {
                    "command": "workbench.action.files.save",
                    "args": []
                },
                {
                    "command": "workbench.action.closeActiveEditor",
                    "args": []
                }
            ]
        },
        {
            "before": [
                "leader",
                "w"
            ],
            "after": [],
            "commands": [
                {
                    "command": "workbench.action.files.save",
                    "args": []
                }
            ]
        }
    ],
    "vim.leader": ",",
    "vim.handleKeys":{
        "<C-a>": false,
        "<C-f>": true,
		"<C-b>": true
    },
    "vim.textwidth": 160,
    "vim.history": 500,
    "workbench.colorTheme": "Visual Studio Dark",
    "typescript.format.insertSpaceAfterFunctionKeywordForAnonymousFunctions": false,
    "eslint.run": "onSave",
    "eslint.autoFixOnSave": true,
    "eslint.validate": [
        "javascript",
        "javascriptreact"
    ],
    "eslint.enable": true,
    "typescript.format.insertSpaceAfterTypeAssertion": true,
	  "typescript.format.insertSpaceAfterOpeningAndBeforeClosingNonemptyBraces": false,
    "editor.tabSize": 2,
    "[markdown]": {
      "files.trimTrailingWhitespace": false
    },
    "editor.renderIndentGuides": true,
    "tslint.autoFixOnSave": true,
    "editor.renderLineHighlight": "all"

}

```

**Comment by bpasero** [comment]
@waitingsong can you clarify what kind of VIM extension you have installed?

**Comment by bpasero** [comment]
@xconverge @jpoon any hints what this command does in your extension that could cause this?

**Comment by xconverge** [comment]
It might be a good idea to disable all plugins, then enable just VSCodeVim. I also see ESlint in the settings which does something onSave... Maybe The vim remap is saving and closing while the eslint extension is doing something in between those 2 commands

**Comment by waitingsong** [comment]
i have disabled all other plugins except 
vscode-language-pack-zh-hans and vim.
have tried under empty folder without package.json, eslint.json, tslibt.json etc, only one file index.ts, still zero after shift zz.

**Comment by waitingsong** [comment]
@xconverge it works. thanks a lot

**Comment by bpasero** [comment]
The issue here was that:
* a file is not dirty
* a file is being saved (e.g. through a command or Ctrl+S)
* a save participant is running (asynchronously, potentially long running)
* a file is closed without waiting for the save to be finished
* the save participant comes back and now wants to save the file
* the file's underlying model has been disposed meanwhile and will now return an empty string as content
* we store an empty string as contents to the file

The fix is now to not save the file if the underlying model has been disposed meanwhile. In other words, it is not possible to save a file and close it right after without waiting for the save to finish. 

This was only an issue when the file was not dirty, because if the file is dirty, we prompt the user before closing the file to either save or revert it.

**Comment by xconverge** [comment]
This was a cool example of how responsive you (the vscode devs) have been with this product. Thanks a lot! A cool example of OSS as well with an efficient issue comment section! (Until this comment ;) )

**Comment by chrmarti** [comment]
Verified I could reproduce without the fix and it no longer reproduces with the fix.

**Comment by joeandaverde** [comment]
What version is the fix in? I’ve had this problem many times on prior and current version (as of 6/8/18) of vscode.

**Comment by chrmarti** [comment]
@joeandaverde This fix is in 1.24.0 and yesterday's insiders build. If you are on these or newer builds, could you open a new bug report?

**Comment by joeandaverde** [comment]
I saw this issue in 1.24.0 and 1.23.1

I thought it was an issue with Prettier. Unfortunately, I am unable to reproduce.

https://github.com/prettier/prettier-vscode/issues/489
